File: csharp/lang/statefulness/static-mutable-collection.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Collections.Immutable;

namespace Example.Orders;

public class OrderRegistry
{
    // ruleid: static-mutable-collection
    private static readonly ConcurrentDictionary<string, Order> SessionCache = new();

    // ruleid: static-mutable-collection
    private static Dictionary<string, Order> sessions = new Dictionary<string, Order>();

    // ruleid: static-mutable-collection
    internal static List<string> PendingIds = new List<string>();

    // ruleid: static-mutable-collection
    private static HashSet<string> seen = new HashSet<string>();

    // ok: static-mutable-collection
    private static readonly ImmutableDictionary<string, string> StatusLabels =
        ImmutableDictionary<string, string>.Empty;

    // ok: static-mutable-collection
    private readonly Dictionary<string, Order> perInstance = new();

    // ok: static-mutable-collection
    private const int MaxRetries = 3;

    public void Register(Order order)
    {
        // ok: static-mutable-collection
        var local = new Dictionary<string, Order> { [order.Id] = order };
        foreach (var kv in local)
        {
            SessionCache[kv.Key] = kv.Value;
        }
    }

    public record Order(string Id);
}

File: java/lang/statefulness/static-mutable-collection.java
package com.example.orders;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

public class OrderRegistry {

  // ruleid: static-mutable-collection
  private static final Map<String, Order> sessionCache = new ConcurrentHashMap<>();

  // ruleid: static-mutable-collection
  static List<String> pendingIds = new ArrayList<>();

  // ruleid: static-mutable-collection
  private static AtomicLong requestCounter = new AtomicLong();

  // ok: static-mutable-collection
  private static final Map<String, String> STATUS_LABELS = new HashMap<>();

  // ok: static-mutable-collection
  private final Map<String, Order> perInstance = new ConcurrentHashMap<>();

  private static final int MAX_RETRIES = 3;

  public void register(Order order) {
    // ok: static-mutable-collection
    Map<String, Order> local = new HashMap<>();
    local.put(order.id(), order);
    sessionCache.putAll(local);
  }

  public record Order(String id) {}
}
